### PR TITLE
Introduce ChainCache and ReadOnlyFilesystemCache

### PR DESCRIPTION
**Pull request by okhoshi**
Following the feedback in symfony/symfony#54384, I’m proposing the two new Cache implementations in Twig directly.

### CODE REVIEW

File: src/Cache/ChainCache.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    public function generateKey(string $name, string $className): string
+    {
+        return $className.'#'.$name;
+    }
+
+    /**
+     * @return string[]
+     */
+    private function splitKey(string $key): array
```
Comment (fabpot) [inline src/Cache/ChainCache.php]: private methods should be move to the end of the class

File: src/Cache/ChainCache.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        foreach ($this->caches as $cache) {
+            if (0 < $timestamp = $cache->getTimestamp($cache->generateKey(...$splitKey))) {
+                return $timestamp;
+            }
+        }
+
+        return 0;
+    }
+}
```
Comment (fabpot) [inline src/Cache/ChainCache.php]: Missing empty line at the end of the file

File: src/Cache/ReadOnlyFilesystemCache.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @author Quentin Devos <quentin@devos.pm>
+ */
+class ReadOnlyFilesystemCache extends FilesystemCache
+{
+    public function write(string $key, string $content): void
+    {
+        // Do nothing with the content, it's a read-only filesystem.
+    }
+}
```
Comment (fabpot) [inline src/Cache/ReadOnlyFilesystemCache.php]: Missing empty line at the end of the file

**Review by fabpot — CHANGES_REQUESTED** [review]
As Twig 3.x is still compatible with PHP 7.2, you need to adjust the code accordingly.

File: src/Cache/ChainCache.php (hunk 1 above)
Comment (okhoshi) [reply src/Cache/ChainCache.php]: Moved

**Comment by okhoshi** [comment]
> As Twig 3.x is still compatible with PHP 7.2, you need to adjust the code accordingly.

Hmm indeed, I missed that. Corrected 👍

File: src/Cache/ChainCache.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ * @author Quentin Devos <quentin@devos.pm>
+ */
+final class ChainCache implements CacheInterface
+{
+    private $caches;
+
+    /**
+     * @param iterable<CacheInterface> $caches The ordered list of caches used to store and fetch cached items
+     */
+    public function __construct(iterable $caches) {
```
Comment (fabpot) [inline src/Cache/ChainCache.php]:
```suggestion
    public function __construct(iterable $caches)
    {
```

**Comment by fabpot** [comment]
Thank you @Okhoshi.

**Pull request merged**